zookeeper-server/src/main/java/org/apache/zookeeper/backup/NFSBackupStorage.java

```
@@ -0,0 +1,324 @@
+            + " does exists or is a directory, could not copy to local storage.");
+      } else {
+        throw new IOException("Cannot open source file in backup storage" + backupFilePath
+            + " , cannot be copied to local storage.");
+      }
+    }
+
+    // Prepare in local storage: create parent dirs if not exist, create the destination file
+    destFile.getParentFile().mkdirs();
+    if (!destFile.createNewFile()) {
+      if (destFile.exists()) {
+        throw new BackupException("A local storage file with the file path " + destFile.getPath()
+            + " already exists, could not create a new one.");
+      } else {
+        throw new IOException("Failed to create a file at destination path: " + destFile.getPath()
+            + ", stopping the streaming.");
+      }
+    }
+
+    // Create output stream on the destination file and start streaming data
+    try {
+      outputStream = new FileOutputStream(destFile);
+
+      byte[] buffer = new byte[1024];
+      int lengthRead;
+      while ((lengthRead = inputStream.read(buffer)) > 0) {
+        outputStream.write(buffer, 0, lengthRead);
+        outputStream.flush();
+      }
+    } catch (FileNotFoundException e) {
+      throw new IOException(
+          "The destination local file " + destFile.getPath() + " could not be opened for write.");
+    } catch (IOException e) {
+      throw new IOException(
+          "IOException is caught during data streaming while copying file " + backupFilePath
+              + " to local storage. The message of the IOException is: " + e.getMessage());
+    } finally {
+      inputStream.close();
+      if (outputStream != null) {
+        outputStream.close();
+      }
+    }
+  }
+
+  @Override
+  public void delete(File fileToDelete) throws IOException {
+    String backupFilePath = constructBackupFilePath(fileToDelete.getPath());
+    try {
+      if (!Files.deleteIfExists(Paths.get(backupFilePath))) {
+        throw new BackupException(
```

Do we need to throw an exception if delete file does not exist?